Error building libmambapy with pybind11 3.0

### Troubleshooting docs

- [x] My problem is not solved in the Troubleshooting docs

### Anaconda default channels

- [x] I do NOT use the Anaconda default channels (pkgs/* etc.)

### How did you install Mamba?

Other (compiled from source on Arch Linux)

### Search tried in issue tracker

utils.cpp

### Latest version of Mamba

- [x] My problem is not solved with the latest version

### Tried in Conda?

Not applicable

### Describe your issue

trying to compile with pybind11 3.0, the below error happens

sadly the line numbers don’t match the main branch in the traceback below, the issue is in lines like this:

https://github.com/mamba-org/mamba/blob/14a94d379a9cbd0f89172bfd320124d947732dec/libmambapy/src/libmambapy/bindings/utils.cpp#L69

### Logs

```log
[129/167] Building CXX object libmambapy/CMakeFiles/bindings.dir/src/libmambapy/bindings/utils.cpp.o
FAILED: libmambapy/CMakeFiles/bindings.dir/src/libmambapy/bindings/utils.cpp.o 
/usr/bin/c++ -DFMT_SHARED -DMAMBA_USE_INSTALL_PREFIX_AS_BASE -DSPDLOG_FMT_EXTERNAL -DSPDLOG_FWRITE_UNLOCKED -Dbindings_EXPORTS -I/build/micromamba/src/mamba-2.3.0/libmambapy/src/libmambapy/bindings -I/build/micromamba/src/mamba-2.3.0/libmamba/include -isystem /usr/include/python3.13 -isystem /usr/share -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fexceptions         -Wp,-D_FORTIFY_SOURCE=3 -Wformat -Werror=format-security         -fstack-clash-protection -fcf-protection         -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -Wp,-D_GLIBCXX_ASSERTIONS -g -ffile-prefix-map=/build/micromamba/src=/usr/src/debug/micromamba -flto=auto -O3 -DNDEBUG -std=c++20 -fPIC -fvisibility=hidden -Wall -Wextra -Wshadow -Wnon-virtual-dtor -Wold-style-cast -Wcast-align -Wunused -Woverloaded-virtual -Wpedantic -Wconversion -Wsign-conversion -Wdouble-promotion -Wformat=2 -Wunreachable-code -Wuninitialized -Wmisleading-indentation -Wduplicated-cond -Wduplicated-branches -Wlogical-op -Wuseless-cast -flto -fno-fat-lto-objects -MD -MT libmambapy/CMakeFiles/bindings.dir/src/libmambapy/bindings/utils.cpp.o -MF libmambapy/CMakeFiles/bindings.dir/src/libmambapy/bindings/utils.cpp.o.d -o libmambapy/CMakeFiles/bindings.dir/src/libmambapy/bindings/utils.cpp.o -c /build/micromamba/src/mamba-2.3.0/libmambapy/src/libmambapy/bindings/utils.cpp
In file included from /usr/include/pybind11/attr.h:14,
                 from /usr/include/pybind11/detail/class.h:12,
                 from /usr/include/pybind11/pybind11.h:12,
                 from /build/micromamba/src/mamba-2.3.0/libmambapy/src/libmambapy/bindings/utils.cpp:11:
/usr/include/pybind11/cast.h: In instantiation of ‘static pybind11::handle pybind11::type::handle_of() [with T = fmt::v11::emphasis]’:
/usr/include/pybind11/detail/../pytypes.h:1583:39:   required from ‘static pybind11::type pybind11::type::of() [with T = fmt::v11::emphasis]’
 1583 |         return type(type::handle_of<T>(), borrowed_t{});
      |                     ~~~~~~~~~~~~~~~~~~^~
/build/micromamba/src/mamba-2.3.0/libmambapy/src/libmambapy/bindings/utils.cpp:33:26:   recursively required by substitution of ‘template<class Func, class Ret> Ret pybind11::init(Func&&) [with Func = fmt::v11::emphasis (*)(const pybind11::str&); Ret = pybind11::detail::initimpl::factory<fmt::v11::emphasis (*)(const pybind11::str&), pybind11::detail::void_type (*)(), fmt::v11::emphasis(const pybind11::str&), pybind11::detail::void_type()>]’
   33 |             .def(py::init(&enum_from_str<fmt::emphasis>));
      |                  ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/build/micromamba/src/mamba-2.3.0/libmambapy/src/libmambapy/bindings/utils.cpp:33:26:   required from here
/usr/include/pybind11/cast.h:2337:89: error: static assertion failed: py::type::of<T> only supports the case where T is a registered C++ types.
 2337 |     static_assert(std::is_base_of<detail::type_caster_generic, detail::make_caster<T>>::value,
      |                                                                                         ^~~~~
/usr/include/pybind11/cast.h:2337:89: note: ‘std::integral_constant<bool, false>::value’ evaluates to false
/usr/include/pybind11/cast.h: In instantiation of ‘static pybind11::handle pybind11::type::handle_of() [with T = fmt::v11::terminal_color]’:
/usr/include/pybind11/detail/../pytypes.h:1583:39:   required from ‘static pybind11::type pybind11::type::of() [with T = fmt::v11::terminal_color]’
 1583 |         return type(type::handle_of<T>(), borrowed_t{});
      |                     ~~~~~~~~~~~~~~~~~~^~
/build/micromamba/src/mamba-2.3.0/libmambapy/src/libmambapy/bindings/utils.cpp:53:26:   recursively required by substitution of ‘template<class Func, class Ret> Ret pybind11::init(Func&&) [with Func = fmt::v11::terminal_color (*)(const pybind11::str&); Ret = pybind11::detail::initimpl::factory<fmt::v11::terminal_color (*)(const pybind11::str&), pybind11::detail::void_type (*)(), fmt::v11::terminal_color(const pybind11::str&), pybind11::detail::void_type()>]’
   53 |             .def(py::init(&enum_from_str<fmt::terminal_color>));
      |                  ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/build/micromamba/src/mamba-2.3.0/libmambapy/src/libmambapy/bindings/utils.cpp:53:26:   required from here
/usr/include/pybind11/cast.h:2337:89: error: static assertion failed: py::type::of<T> only supports the case where T is a registered C++ types.
 2337 |     static_assert(std::is_base_of<detail::type_caster_generic, detail::make_caster<T>>::value,
      |                                                                                         ^~~~~
/usr/include/pybind11/cast.h:2337:89: note: ‘std::integral_constant<bool, false>::value’ evaluates to false
```

--- a/dev/environment-dev.yml
+++ b/dev/environment-dev.yml
@@ -53,7 +53,7 @@ dependencies:
   - scikit-build-core
   # libmambapy dependencies
   - python
-  - pybind11<3.0.0
+  - pybind11>=3.0.0
   # libmambapy-stubs build dependencies
   - mypy # For stubgen
   - setuptools

--- a/libmambapy/bindings/bind_utils.hpp
+++ b/libmambapy/bindings/bind_utils.hpp
@@ -14,11 +14,45 @@
 
 namespace mambapy
 {
-    template <typename Enum>
-    auto enum_from_str(const pybind11::str& name)
+    /**
+     * Bind enum values and add a converting string constructor.
+     *
+     * This functions appeared during the migration from pybind 2 to 3.
+     * Pybind 3 introduced py::native_enum, and made changes to previous py::enum.
+     * At this point, libmambapy was relying on the fact that one could create an enum
+     * from a string, using a constructor, and an implicit conversion to pass function enum
+     * parameters as strings (Python catnip).
+     * This function was added to avoid breaking that contract.
+     * ``pybind11::enum_`` was kept because ``pybind11::native_enum`` does not support
+     * to this day constructors needed for implicit conversion.
+     *
+     * Perhaps ``pybind11::native_enum`` will support string implicit conversion in the future,
+     * otherwise the implicit conversion could be broken in a major release.
+     */
+    template <typename Enum, std::size_t N, typename PyEnum>
+    auto make_str_enum(PyEnum&& pyenum, std::array<std::pair<const char*, Enum>, N> name_values)
     {
-        auto pyenum = pybind11::type::of<Enum>();
-        return pyenum.attr("__members__")[name].template cast<Enum>();
+        for (const auto& [name, val] : name_values)
+        {
+            pyenum.value(name, val);
+        }
+
+        pyenum.def(pybind11::init(
+            [name_values](std::string_view s)
+            {
+                for (const auto& [name, val] : name_values)
+                {
+                    if (name == s)
+                    {
+                        return val;
+                    }
+                }
+                throw pybind11::key_error(std::format("No member named {}", s));
+            }
+        ));
+
+        pybind11::implicitly_convertible<pybind11::str, Enum>();
+        return std::forward<PyEnum>(pyenum);
     }
 
     template <typename T>

--- a/libmambapy/bindings/bind_utils.hpp
+++ b/libmambapy/bindings/bind_utils.hpp
@@ -7,6 +7,7 @@
 #ifndef LIBMAMBAPY_BIND_UTILS_HPP
 #define LIBMAMBAPY_BIND_UTILS_HPP
 
+#include <format>
 #include <memory>
 
 #include <pybind11/operators.h>

--- a/libmambapy/bindings/legacy.cpp
+++ b/libmambapy/bindings/legacy.cpp
@@ -386,11 +386,11 @@ bind_submodule_impl(pybind11::module_ m)
     m.attr("SOLVER_SETNAME") = global_solver_job_v2_migrator;
     m.attr("SOLVER_SETMASK") = global_solver_job_v2_migrator;
 
-    enum struct SolverRuleinfoV2Migrator
+    struct SolverRuleinfoV2Migrator
     {
     };
 
-    py::enum_<SolverRuleinfoV2Migrator>(m, "SolverRuleinfo")
+    py::class_<SolverRuleinfoV2Migrator>(m, "SolverRuleinfo")
         .def(py::init(
             [](py::args, py::kwargs) -> SolverRuleinfoV2Migrator
             {
@@ -398,11 +398,11 @@ bind_submodule_impl(pybind11::module_ m)
             }
         ));
 
-    enum struct SolverV2Migrator
+    struct SolverV2Migrator
     {
     };
 
-    py::enum_<SolverV2Migrator>(m, "Solver")
+    py::class_<SolverV2Migrator>(m, "Solver")
         .def(py::init(
             [](py::args, py::kwargs) -> SolverV2Migrator
             {

--- a/libmambapy/bindings/legacy.cpp
+++ b/libmambapy/bindings/legacy.cpp
@@ -1484,22 +1484,26 @@ bind_submodule_impl(pybind11::module_ m)
 
     ////////////////////////////////////////////
 
-    py::enum_<QueryType>(m, "QueryType")
-        .value("Search", QueryType::Search)
-        .value("Depends", QueryType::Depends)
-        .value("WhoNeeds", QueryType::WhoNeeds)
-        .def(py::init(&mambapy::enum_from_str<QueryType>))
+    mambapy::make_str_enum(
+        py::enum_<QueryType>(m, "QueryType"),
+        std::array{
+            std::pair{ "Search", QueryType::Search },
+            std::pair{ "Depends", QueryType::Depends },
+            std::pair{ "WhoNeeds", QueryType::WhoNeeds },
+        }
+    )
         .def_static("parse", &query_type_parse);
-    py::implicitly_convertible<py::str, QueryType>();
 
-    py::enum_<QueryResultFormat>(m, "QueryResultFormat")
-        .value("Json", QueryResultFormat::Json)
-        .value("Tree", QueryResultFormat::Tree)
-        .value("Table", QueryResultFormat::Table)
-        .value("Pretty", QueryResultFormat::Pretty)
-        .value("RecursiveTable", QueryResultFormat::RecursiveTable)
-        .def(py::init(&mambapy::enum_from_str<QueryResultFormat>));
-    py::implicitly_convertible<py::str, QueryType>();
+    mambapy::make_str_enum(
+        py::enum_<QueryResultFormat>(m, "QueryResultFormat"),
+        std::array{
+            std::pair{ "Json", QueryResultFormat::Json },
+            std::pair{ "Tree", QueryResultFormat::Tree },
+            std::pair{ "Table", QueryResultFormat::Table },
+            std::pair{ "Pretty", QueryResultFormat::Pretty },
+            std::pair{ "RecursiveTable", QueryResultFormat::RecursiveTable },
+        }
+    );
 
     py::class_<QueryResult>(m, "QueryResult")
         .def_property_readonly("type", &QueryResult::type)

--- a/libmambapy/bindings/solver_libsolv.cpp
+++ b/libmambapy/bindings/solver_libsolv.cpp
@@ -29,46 +29,60 @@ namespace mambapy
         using namespace mamba;
         using namespace mamba::solver::libsolv;
 
-        py::enum_<RepodataParser>(m, "RepodataParser")
-            .value("Mamba", RepodataParser::Mamba)
-            .value("Libsolv", RepodataParser::Libsolv)
-            .def(py::init(&enum_from_str<RepodataParser>));
-        py::implicitly_convertible<py::str, RepodataParser>();
+        make_str_enum(
+            py::enum_<RepodataParser>(m, "RepodataParser"),
+            std::array{
+                std::pair{ "Mamba", RepodataParser::Mamba },
+                std::pair{ "Libsolv", RepodataParser::Libsolv },
+            }
+        );
 
-        py::enum_<MatchSpecParser>(m, "MatchSpecParser")
-            .value("Mixed", MatchSpecParser::Mixed)
-            .value("Mamba", MatchSpecParser::Mamba)
-            .value("Libsolv", MatchSpecParser::Libsolv)
-            .def(py::init(&enum_from_str<MatchSpecParser>));
-        py::implicitly_convertible<py::str, MatchSpecParser>();
+        make_str_enum(
+            py::enum_<MatchSpecParser>(m, "MatchSpecParser"),
+            std::array{
+                std::pair{ "Mixed", MatchSpecParser::Mixed },
+                std::pair{ "Mamba", MatchSpecParser::Mamba },
+                std::pair{ "Libsolv", MatchSpecParser::Libsolv },
+            }
+        );
 
-        py::enum_<PipAsPythonDependency>(m, "PipAsPythonDependency")
-            .value("No", PipAsPythonDependency::No)
-            .value("Yes", PipAsPythonDependency::Yes)
-            .def(py::init([](bool val) { return static_cast<PipAsPythonDependency>(val); }));
+        make_str_enum(
+            py::enum_<PipAsPythonDependency>(m, "PipAsPythonDependency"),
+            std::array{
+                std::pair{ "No", PipAsPythonDependency::No },
+                std::pair{ "Yes", PipAsPythonDependency::Yes },
+            }
+        );
         py::implicitly_convertible<py::bool_, PipAsPythonDependency>();
 
-        py::enum_<PackageTypes>(m, "PackageTypes")
-            .value("CondaOnly", PackageTypes::CondaOnly)
-            .value("TarBz2Only", PackageTypes::TarBz2Only)
-            .value("CondaAndTarBz2", PackageTypes::CondaAndTarBz2)
-            .value("CondaOrElseTarBz2", PackageTypes::CondaOrElseTarBz2)
-            .def(py::init(&enum_from_str<PackageTypes>));
-        py::implicitly_convertible<py::str, PackageTypes>();
+        make_str_enum(
+            py::enum_<PackageTypes>(m, "PackageTypes"),
+            std::array{
+                std::pair{ "CondaOnly", PackageTypes::CondaOnly },
+                std::pair{ "TarBz2Only", PackageTypes::TarBz2Only },
+                std::pair{ "CondaAndTarBz2", PackageTypes::CondaAndTarBz2 },
+                std::pair{ "CondaOrElseTarBz2", PackageTypes::CondaOrElseTarBz2 },
+            }
+        );
 
-        py::enum_<VerifyPackages>(m, "VerifyPackages")
-            .value("No", VerifyPackages::No)
-            .value("Yes", VerifyPackages::Yes)
-            .def(py::init([](bool val) { return static_cast<VerifyPackages>(val); }));
+        make_str_enum(
+            py::enum_<VerifyPackages>(m, "VerifyPackages"),
+            std::array{
+                std::pair{ "No", VerifyPackages::No },
+                std::pair{ "Yes", VerifyPackages::Yes },
+            }
+        );
         py::implicitly_convertible<py::bool_, VerifyPackages>();
 
-        py::enum_<LogLevel>(m, "LogLevel")
-            .value("Debug", LogLevel::Debug)
-            .value("Warning", LogLevel::Warning)
-            .value("Error", LogLevel::Error)
-            .value("Fatal", LogLevel::Fatal)
-            .def(py::init(&enum_from_str<LogLevel>));
-        py::implicitly_convertible<py::bool_, LogLevel>();
+        make_str_enum(
+            py::enum_<LogLevel>(m, "LogLevel"),
+            std::array{
+                std::pair{ "Debug", LogLevel::Debug },
+                std::pair{ "Warning", LogLevel::Warning },
+                std::pair{ "Error", LogLevel::Error },
+                std::pair{ "Fatal", LogLevel::Fatal },
+            }
+        );
 
         py::class_<Priorities>(m, "Priorities")
             .def(

--- a/libmambapy/bindings/specs.cpp
+++ b/libmambapy/bindings/specs.cpp
@@ -63,47 +63,53 @@ namespace mambapy
             [](const mamba::fs::u8path& p) { return strip_archive_extension(p); }
         );
 
-        py::enum_<KnownPlatform>(m, "KnownPlatform")
-            .value("noarch", KnownPlatform::noarch)
-            .value("linux_32", KnownPlatform::linux_32)
-            .value("linux_64", KnownPlatform::linux_64)
-            .value("linux_armv6l", KnownPlatform::linux_armv6l)
-            .value("linux_armv7l", KnownPlatform::linux_armv7l)
-            .value("linux_aarch64", KnownPlatform::linux_aarch64)
-            .value("linux_ppc64le", KnownPlatform::linux_ppc64le)
-            .value("linux_ppc64", KnownPlatform::linux_ppc64)
-            .value("linux_s390x", KnownPlatform::linux_s390x)
-            .value("linux_riscv32", KnownPlatform::linux_riscv32)
-            .value("linux_riscv64", KnownPlatform::linux_riscv64)
-            .value("osx_64", KnownPlatform::osx_64)
-            .value("osx_arm64", KnownPlatform::osx_arm64)
-            .value("win_32", KnownPlatform::win_32)
-            .value("win_64", KnownPlatform::win_64)
-            .value("win_arm64", KnownPlatform::win_arm64)
-            .value("zos_z", KnownPlatform::zos_z)
-            .def(py::init(&enum_from_str<KnownPlatform>))
+        make_str_enum(
+            py::enum_<KnownPlatform>(m, "KnownPlatform"),
+            std::array{
+                std::pair{ "noarch", KnownPlatform::noarch },
+                std::pair{ "linux_32", KnownPlatform::linux_32 },
+                std::pair{ "linux_64", KnownPlatform::linux_64 },
+                std::pair{ "linux_armv6l", KnownPlatform::linux_armv6l },
+                std::pair{ "linux_armv7l", KnownPlatform::linux_armv7l },
+                std::pair{ "linux_aarch64", KnownPlatform::linux_aarch64 },
+                std::pair{ "linux_ppc64le", KnownPlatform::linux_ppc64le },
+                std::pair{ "linux_ppc64", KnownPlatform::linux_ppc64 },
+                std::pair{ "linux_s390x", KnownPlatform::linux_s390x },
+                std::pair{ "linux_riscv32", KnownPlatform::linux_riscv32 },
+                std::pair{ "linux_riscv64", KnownPlatform::linux_riscv64 },
+                std::pair{ "osx_64", KnownPlatform::osx_64 },
+                std::pair{ "osx_arm64", KnownPlatform::osx_arm64 },
+                std::pair{ "win_32", KnownPlatform::win_32 },
+                std::pair{ "win_64", KnownPlatform::win_64 },
+                std::pair{ "win_arm64", KnownPlatform::win_arm64 },
+                std::pair{ "zos_z", KnownPlatform::zos_z },
+            }
+        )
             .def_static("parse", &platform_parse)
             .def_static("count", &known_platforms_count)
             .def_static("build_platform", &build_platform);
-        py::implicitly_convertible<py::str, KnownPlatform>();
 
-        py::enum_<NoArchType>(m, "NoArchType")
-            .value("No", NoArchType::No)
-            .value("Generic", NoArchType::Generic)
-            .value("Python", NoArchType::Python)
-            .def(py::init(&enum_from_str<NoArchType>))
+        make_str_enum(
+            py::enum_<NoArchType>(m, "NoArchType"),
+            std::array{
+                std::pair{ "No", NoArchType::No },
+                std::pair{ "Generic", NoArchType::Generic },
+                std::pair{ "Python", NoArchType::Python },
+            }
+        )
             .def_static("parse", &noarch_parse)
             .def_static("count", &known_noarch_count);
-        py::implicitly_convertible<py::str, NoArchType>();
 
         auto py_conda_url = py::class_<CondaURL>(m, "CondaURL");
 
-        py::enum_<CondaURL::Credentials>(py_conda_url, "Credentials")
-            .value("Hide", CondaURL::Credentials::Hide)
-            .value("Show", CondaURL::Credentials::Show)
-            .value("Remove", CondaURL::Credentials::Remove)
-            .def(py::init(&enum_from_str<CondaURL::Credentials>));
-        py::implicitly_convertible<py::str, CondaURL::Credentials>();
+        make_str_enum(
+            py::enum_<CondaURL::Credentials>(py_conda_url, "Credentials"),
+            std::array{
+                std::pair{ "Hide", CondaURL::Credentials::Hide },
+                std::pair{ "Show", CondaURL::Credentials::Show },
+                std::pair{ "Remove", CondaURL::Credentials::Remove },
+            }
+        );
 
         py_conda_url  //
             .def_static("parse", &CondaURL::parse)
@@ -315,15 +321,17 @@ namespace mambapy
 
         auto py_unresolved_channel = py::class_<UnresolvedChannel>(m, "UnresolvedChannel");
 
-        py::enum_<UnresolvedChannel::Type>(py_unresolved_channel, "Type")
-            .value("URL", UnresolvedChannel::Type::URL)
-            .value("PackageURL", UnresolvedChannel::Type::PackageURL)
-            .value("Path", UnresolvedChannel::Type::Path)
-            .value("PackagePath", UnresolvedChannel::Type::PackagePath)
-            .value("Name", UnresolvedChannel::Type::Name)
-            .value("Unknown", UnresolvedChannel::Type::Unknown)
-            .def(py::init(&enum_from_str<UnresolvedChannel::Type>));
-        py::implicitly_convertible<py::str, UnresolvedChannel::Type>();
+        make_str_enum(
+            py::enum_<UnresolvedChannel::Type>(py_unresolved_channel, "Type"),
+            std::array{
+                std::pair{ "URL", UnresolvedChannel::Type::URL },
+                std::pair{ "PackageURL", UnresolvedChannel::Type::PackageURL },
+                std::pair{ "Path", UnresolvedChannel::Type::Path },
+                std::pair{ "PackagePath", UnresolvedChannel::Type::PackagePath },
+                std::pair{ "Name", UnresolvedChannel::Type::Name },
+                std::pair{ "Unknown", UnresolvedChannel::Type::Unknown },
+            }
+        );
 
         py_unresolved_channel  //
             .def_static("parse", UnresolvedChannel::parse)
@@ -436,12 +444,14 @@ namespace mambapy
             .def("__copy__", &copy<BasicHTTPAuthentication>)
             .def("__deepcopy__", &deepcopy<BasicHTTPAuthentication>, py::arg("memo"));
 
-        py::enum_<Channel::Match>(py_channel, "Match")
-            .value("No", Channel::Match::No)
-            .value("InOtherPlatform", Channel::Match::InOtherPlatform)
-            .value("Full", Channel::Match::Full)
-            .def(py::init(&enum_from_str<Channel::Match>));
-        py::implicitly_convertible<py::str, Channel::Match>();
+        make_str_enum(
+            py::enum_<Channel::Match>(py_channel, "Match"),
+            std::array{
+                std::pair{ "No", Channel::Match::No },
+                std::pair{ "InOtherPlatform", Channel::Match::InOtherPlatform },
+                std::pair{ "Full", Channel::Match::Full },
+            }
+        );
 
         py_channel  //
             .def_property_readonly_static(

--- a/libmambapy/bindings/utils.cpp
+++ b/libmambapy/bindings/utils.cpp
@@ -57,37 +57,41 @@ namespace mambapy
     {
         namespace py = pybind11;
 
-        py::enum_<fmt::emphasis>(m, "TextEmphasis")
-            .value("Bold", fmt::emphasis::bold)
-            .value("Faint", fmt::emphasis::faint)
-            .value("Italic", fmt::emphasis::italic)
-            .value("Underline", fmt::emphasis::underline)
-            .value("Blink", fmt::emphasis::blink)
-            .value("Reverse", fmt::emphasis::reverse)
-            .value("Conceal", fmt::emphasis::conceal)
-            .value("Strikethrough", fmt::emphasis::strikethrough)
-            .def(py::init(&enum_from_str<fmt::emphasis>));
-        py::implicitly_convertible<py::str, fmt::emphasis>();
+        make_str_enum(
+            py::enum_<fmt::emphasis>(m, "TextEmphasis"),
+            std::array{
+                std::pair{ "Bold", fmt::emphasis::bold },
+                std::pair{ "Faint", fmt::emphasis::faint },
+                std::pair{ "Italic", fmt::emphasis::italic },
+                std::pair{ "Underline", fmt::emphasis::underline },
+                std::pair{ "Blink", fmt::emphasis::blink },
+                std::pair{ "Reverse", fmt::emphasis::reverse },
+                std::pair{ "Conceal", fmt::emphasis::conceal },
+                std::pair{ "Strikethrough", fmt::emphasis::strikethrough },
+            }
+        );
 
-        py::enum_<fmt::terminal_color>(m, "TextTerminalColor")
-            .value("Black", fmt::terminal_color::black)
-            .value("Red", fmt::terminal_color::red)
-            .value("Green", fmt::terminal_color::green)
-            .value("Yellow", fmt::terminal_color::yellow)
-            .value("Blue", fmt::terminal_color::blue)
-            .value("Magenta", fmt::terminal_color::magenta)
-            .value("Cyan", fmt::terminal_color::cyan)
-            .value("White", fmt::terminal_color::white)
-            .value("BrightBlack", fmt::terminal_color::bright_black)
-            .value("BrightRed", fmt::terminal_color::bright_red)
-            .value("BrightGreen", fmt::terminal_color::bright_green)
-            .value("BrightYellow", fmt::terminal_color::bright_yellow)
-            .value("BrightBlue", fmt::terminal_color::bright_blue)
-            .value("BrightMagenta", fmt::terminal_color::bright_magenta)
-            .value("BrightCyan", fmt::terminal_color::bright_cyan)
-            .value("BrightWhite", fmt::terminal_color::bright_white)
-            .def(py::init(&enum_from_str<fmt::terminal_color>));
-        py::implicitly_convertible<py::str, fmt::terminal_color>();
+        make_str_enum(
+            py::enum_<fmt::terminal_color>(m, "TextTerminalColor"),
+            std::array{
+                std::pair{ "Black", fmt::terminal_color::black },
+                std::pair{ "Red", fmt::terminal_color::red },
+                std::pair{ "Green", fmt::terminal_color::green },
+                std::pair{ "Yellow", fmt::terminal_color::yellow },
+                std::pair{ "Blue", fmt::terminal_color::blue },
+                std::pair{ "Magenta", fmt::terminal_color::magenta },
+                std::pair{ "Cyan", fmt::terminal_color::cyan },
+                std::pair{ "White", fmt::terminal_color::white },
+                std::pair{ "BrightBlack", fmt::terminal_color::bright_black },
+                std::pair{ "BrightRed", fmt::terminal_color::bright_red },
+                std::pair{ "BrightGreen", fmt::terminal_color::bright_green },
+                std::pair{ "BrightYellow", fmt::terminal_color::bright_yellow },
+                std::pair{ "BrightBlue", fmt::terminal_color::bright_blue },
+                std::pair{ "BrightMagenta", fmt::terminal_color::bright_magenta },
+                std::pair{ "BrightCyan", fmt::terminal_color::bright_cyan },
+                std::pair{ "BrightWhite", fmt::terminal_color::bright_white },
+            }
+        );
 
         py::class_<fmt::rgb>(m, "TextRGBColor")
             .def(py::init())

